# I. Import Necessary Libraries

In [10]:
import pandas as pd
import numpy as np
import os 
import matplotlib.pyplot as plt
import seaborn as sns

In [23]:
df = pd.read_csv("data.csv")
df




City  age_pyramid_total_nyc  \
0  New York city, New York                8550405   
1  New York city, New York                8537673   
2  New York city, New York                8622698   
3  New York city, New York                8398748   
4  New York city, New York                8336817   
5  New York city, New York                8467513   

   median_household_income_nyc  median_personal_earnings_nyc  no_hs_nyc  \
0                        55752                         36087  19.097573   
1                        58856                         36871  18.456166   
2                        60879                         38430  18.114607   
3                        63799                         40932  17.287630   
4                        69407                         42326  16.779843   
5                        67997                         43133  16.133492   

   at_least_hs_nyc  complete_hs_somecollege_nyc  complete_bach_nyc  \
0        80.902427                    44.058038          21.885647   
1        81.543834                    44.523201          21.543797   
2        81.885393                    44.563906          21.631749   
3        82.712370                    43.728064          22.835438   
4        83.220157                    44.000053          22.727674   
5        83.866508                    42.906229          23.516056   

   at_least_bachelors_nyc  grad_degree_nyc  ...  veterans_poverty_nyc  \
0               36.844389        14.958742  ...                 17862   
1               37.020633        15.476836  ...                 14778   
2               37.321486        15.689738  ...                 13927   
3               38.984307        16.148868  ...                 15986   
4               39.220104        16.492430  ...                 16280   
5               40.960278        17.444222  ...                 15750   

   veterans_poverty_percent_nyc  veterans_poverty_disabled_nyc  \
0                     11.379897                           7520   
1                     10.027073                           6719   
2                      9.233880                           4623   
3                     11.129367                           6252   
4                     11.789583                           7352   
5                     11.808101                           5885   

   veterans_poverty_disabled_percent_nyc  vet_total_nyc_18_64  \
0                               4.790999                69775   
1                               4.558932                70710   
2                               3.065142                68674   
3                               4.352609                64621   
4                               5.324141                64353   
5                               4.412106                60375   

   veterans_unemployed_nyc  veterans_unemployed_percent_nyc  \
0                     3139                         4.498746   
1                     3362                         4.754632   
2                     3958                         5.763462   
3                     2537                         3.925968   
4                     2559                         3.976505   
5                     3588                         5.942857   

   veterans_disabled_nyc  veterans_disabled_percent_nyc  year  
0                  24730                      15.755506  2015  
1                  25358                      17.205746  2016  
2                  20653                      13.693353  2017  
3                  21320                      14.842869  2018  
4                  20634                      14.942645  2019  
5                  22002                      16.495355  2021  

[6 rows x 199 columns]

In [24]:
# DROP THE CITY COLUMN, It is unnecessary for our correlation
df = df.drop('City', axis=1)

# Replace 'Year' with the actual column name containing the years in your DataFrame
df.set_index('year', inplace=True)

df

age_pyramid_total_nyc  median_household_income_nyc  \
year                                                       
2015                8550405                        55752   
2016                8537673                        58856   
2017                8622698                        60879   
2018                8398748                        63799   
2019                8336817                        69407   
2021                8467513                        67997   

      median_personal_earnings_nyc  no_hs_nyc  at_least_hs_nyc  \
year                                                             
2015                         36087  19.097573        80.902427   
2016                         36871  18.456166        81.543834   
2017                         38430  18.114607        81.885393   
2018                         40932  17.287630        82.712370   
2019                         42326  16.779843        83.220157   
2021                         43133  16.133492        83.866508   

      complete_hs_somecollege_nyc  complete_bach_nyc  at_least_bachelors_nyc  \
year                                                                           
2015                    44.058038          21.885647               36.844389   
2016                    44.523201          21.543797               37.020633   
2017                    44.563906          21.631749               37.321486   
2018                    43.728064          22.835438               38.984307   
2019                    44.000053          22.727674               39.220104   
2021                    42.906229          23.516056               40.960278   

      grad_degree_nyc  latino_pop_nyc  ...  veterans_med_income_nyc  \
year                                   ...                            
2015        14.958742       29.064413  ...                    39118   
2016        15.476836       29.154197  ...                    40942   
2017        15.689738       29.195375  ...                    41487   
2018        16.148868       29.164478  ...                    38101   
2019        16.492430       29.070903  ...                    44748   
2021        17.444222       29.100871  ...                    43905   

      veterans_poverty_nyc  veterans_poverty_percent_nyc  \
year                                                       
2015                 17862                     11.379897   
2016                 14778                     10.027073   
2017                 13927                      9.233880   
2018                 15986                     11.129367   
2019                 16280                     11.789583   
2021                 15750                     11.808101   

      veterans_poverty_disabled_nyc  veterans_poverty_disabled_percent_nyc  \
year                                                                         
2015                           7520                               4.790999   
2016                           6719                               4.558932   
2017                           4623                               3.065142   
2018                           6252                               4.352609   
2019                           7352                               5.324141   
2021                           5885                               4.412106   

      vet_total_nyc_18_64  veterans_unemployed_nyc  \
year                                                 
2015                69775                     3139   
2016                70710                     3362   
2017                68674                     3958   
2018                64621                     2537   
2019                64353                     2559   
2021                60375                     3588   

      veterans_unemployed_percent_nyc  veterans_disabled_nyc  \
year                                                           
2015                         4.498746                  24730   
2016                         4.754632                  

In [29]:
# II. Produce Correlation Matrix
correlation_matrix = df.corr()
correlation_matrix


age_pyramid_total_nyc  \
age_pyramid_total_nyc                         1.000000   
median_household_income_nyc                  -0.746539   
median_personal_earnings_nyc                 -0.721950   
no_hs_nyc                                     0.662630   
at_least_hs_nyc                              -0.662630   
...                                                ...   
vet_total_nyc_18_64                           0.631425   
veterans_unemployed_nyc                       0.850547   
veterans_unemployed_percent_nyc               0.628440   
veterans_disabled_nyc                         0.390758   
veterans_disabled_percent_nyc                -0.030474   

                                 median_household_income_nyc  \
age_pyramid_total_nyc                              -0.746539   
median_household_income_nyc                         1.000000   
median_personal_earnings_nyc                        0.972589   
no_hs_nyc                                          -0.961426   
at_least_hs_nyc                                     0.961426   
...                                                      ...   
vet_total_nyc_18_64                                -0.871934   
veterans_unemployed_nyc                            -0.302724   
veterans_unemployed_percent_nyc                     0.003386   
veterans_disabled_nyc                              -0.716601   
veterans_disabled_percent_nyc                      -0.126211   

                                 median_personal_earnings_nyc  no_hs_nyc  \
age_pyramid_total_nyc                               -0.721950   0.662630   
median_household_income_nyc                          0.972589  -0.961426   
median_personal_earnings_nyc                         1.000000  -0.988286   
no_hs_nyc                                           -0.988286   1.000000   
at_least_hs_nyc                                      0.988286  -1.000000   
...                                                       ...        ...   
vet_total_nyc_18_64                                 -0.955594   0.953129   
veterans_unemployed_nyc                             -0.286138   0.192194   
veterans_unemployed_percent_nyc                      0.053649  -0.150513   
veterans_disabled_nyc                               -0.718335   0.648040   
veterans_disabled_percent_nyc                       -0.125271   0.016984   

                                 at_least_hs_nyc  complete_hs_somecollege_nyc  \
age_pyramid_total_nyc                  -0.662630                     0.471650   
median_household_income_nyc             0.961426                    -0.588893   
median_personal_earnings_nyc            0.988286                    -0.729969   
no_hs_nyc                              -1.000000                     0.746818   
at_least_hs_nyc                         1.000000                    -0.746818   
...                                          ...                          ...   
vet_total_nyc_18_64                    -0.953129                     0.893675   
veterans_unemployed_nyc                -0.192194                     0.152005   
veterans_unemployed_percent_nyc         0.150513                    -0.192022   
veterans_disabled_nyc                  -0.648040                     0.226652   
veterans_disabled_percent_nyc          -0.016984                    -0.266105   

                                 complete_bach_nyc  at_least_bachelors_nyc  \
age_pyramid_total_nyc                    -0.695634               -0.632688   
median_household_income_nyc               0.805786                0.881985   
median_personal_earnings_nyc              0.905428                0.953633   
no_hs_nyc                                -0.896998               -0.968028   
at_least_hs_nyc                           0.896998                0.968028   
...                                            ...                     ...   
vet_total_nyc_18_64                      -0.978392               -0.991252   
veterans_unemployed_nyc                  -0.338704       


Correlation Coefficient Values: Each cell in the correlation matrix represents the correlation coefficient between two variables. The values range from -1 to 1:

- +1 indicates a perfect positive linear relationship.
- -1 indicates a perfect negative linear relationship.
- 0 indicates no linear relationship.

High Correlation: Generally, a correlation coefficient above 0.7 or below -0.7 is considered strong, suggesting a significant linear relationship between the two variables.

Diagonal Line: The diagonal from the top left to the bottom right shows the correlation of each variable with itself, which is always 1 (a perfect positive correlation).

To determine which variables are most strongly correlated, you'll want to look for the numbers that are closest to 1 or -1. It's important to consider both positive and negative values.

For example, in your matrix:

The variable median_household_income_nyc has a strong positive correlation with median_personal_earnings_nyc (0.975289).
Conversely, median_household_income_nyc has a strong negative correlation with no_hs_nyc (-0.961426), meaning that as the median household income increases, the proportion of people without a high school diploma decreases (or vice versa).

In [17]:
def get_top_correlated_variables(correlation_matrix, target_variable, n=5):
    if target_variable not in correlation_matrix:
        raise ValueError(f"{target_variable} is not a column in the correlation matrix")
    
    # Get the absolute values of correlations for the target variable
    target_correlations = correlation_matrix[target_variable].abs()
    
    # Drop the self-correlation of the variable
    target_correlations = target_correlations.drop(target_variable, errors='ignore')
    
    # Get the top n most correlated variables
    top_correlated_variables = target_correlations.sort_values(ascending=False).head(n)
    
    return top_correlated_variables
print(get_top_correlated_variables(correlation_matrix, "veterans_disabled_percent_nyc", 5))


west_asia_nyc                          0.888997
nec_asia_nyc                           0.786660
veterans_disabled_nyc                  0.768516
perc_age_pyramid_female_5_to_14_nyc    0.749129
perc_age_pyramid_5_to_14_nyc           0.736947
Name: veterans_disabled_percent_nyc, dtype: float64
